# Assignment Chapter 2 - MACHINE LEARNING [Case #1]
Startup Campus, Indonesia - `Artificial Intelligence (AI)` (Batch 7)
* Dataset: SC_HW1_bank_data
* Libraries: Pandas, Numpy, Scikit-learn
* Objective: Classification with Supervised Machine Learning Models

`PERSYARATAN` Semua modul (termasuk versi yang sesuai) sudah di-install dengan benar.
<br>`CARA PENGERJAAN` Lengkapi baris kode yang ditandai dengan **#TODO**.
<br>`TARGET PORTFOLIO` Peserta mampu membandingkan akurasi klasifikasi dari berbagai model *supervised learning*.

### Import Library

In [ ]:
import pandas as pd
import numpy as np

from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split as tts
from sklearn.model_selection import GridSearchCV

from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix

### Read Dataset

In [ ]:
df = pd.read_csv('https://raw.githubusercontent.com/Rietaros/kampus_merdeka/main/SC_HW1_bank_data.csv')
df.columns
df

RowNumber  CustomerId Geography  Gender  Age  Tenure    Balance  \
0             1    15634602    France  Female   42       2       0.00   
1             2    15647311     Spain  Female   41       1   83807.86   
2             3    15619304    France  Female   42       8  159660.80   
3             4    15701354    France  Female   39       1       0.00   
4             5    15737888     Spain  Female   43       2  125510.82   
...         ...         ...       ...     ...  ...     ...        ...   
9995       9996    15606229    France    Male   39       5       0.00   
9996       9997    15569892    France    Male   35      10   57369.61   
9997       9998    15584532    France  Female   36       7       0.00   
9998       9999    15682355   Germany    Male   42       3   75075.31   
9999      10000    15628319    France  Female   28       4  130142.79   

      NumOfProducts  HasCrCard  IsActiveMember  EstimatedSalary  Exited  
0                 1          1               1        101348.88       1  
1                 1          0               1        112542.58       0  
2                 3          1               0        113931.57       1  
3                 2          0               0         93826.63       0  
4                 1          1               1         79084.10       0  
...             ...        ...             ...              ...     ...  
9995              2          1               0         96270.64       0  
9996              1          1               1        101699.77       0  
9997              1          0               1         42085.58       1  
9998              2          1               0         92888.52       1  
9999              1          1               0         38190.78       0  

[10000 rows x 12 columns]

In [ ]:
# TODO: Hilangkan kolom yang tidak relevan untuk pemodelan
df = df[['Geography', 'Gender', 'Age', 'Tenure', 'Balance',
         'NumOfProducts', 'HasCrCard', 'IsActiveMember',
         'EstimatedSalary', 'Exited']].copy()
df.head()

Geography  Gender  Age  Tenure    Balance  NumOfProducts  HasCrCard  \
0    France  Female   42       2       0.00              1          1   
1     Spain  Female   41       1   83807.86              1          0   
2    France  Female   42       8  159660.80              3          1   
3    France  Female   39       1       0.00              2          0   
4     Spain  Female   43       2  125510.82              1          1   

   IsActiveMember  EstimatedSalary  Exited  
0               1        101348.88       1  
1               1        112542.58       0  
2               0        113931.57       1  
3               0         93826.63       0  
4               1         79084.10       0

### Preprocessing

In [ ]:
# TODO: Lakukan One-Hot Encoder untuk data kategorikal, dengan fungsi pd.get_dummies
df = pd.get_dummies(df, columns=['Gender', 'Geography'], drop_first=False)
df = df.astype(int)
df.head()

Age  Tenure  Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0   42       2        0              1          1               1   
1   41       1    83807              1          0               1   
2   42       8   159660              3          1               0   
3   39       1        0              2          0               0   
4   43       2   125510              1          1               1   

   EstimatedSalary  Exited  Gender_Female  Gender_Male  Geography_France  \
0           101348       1              1            0                 1   
1           112542       0              1            0                 0   
2           113931       1              1            0                 1   
3            93826       0              1            0                 1   
4            79084       0              1            0                 0   

   Geography_Germany  Geography_Spain  
0                  0                0  
1                  0                1  
2                  0                0  
3                  0                0  
4                  0                1

In [ ]:
# Pisahkan fitur (X) dengan target (y)
X = df.drop(columns=['Exited'])
y = df['Exited']

print("Fitur (X):")
print(X.head())

print("\nTarget (y):")
print(y.head())

Fitur (X):
   Age  Tenure  Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0   42       2        0              1          1               1   
1   41       1    83807              1          0               1   
2   42       8   159660              3          1               0   
3   39       1        0              2          0               0   
4   43       2   125510              1          1               1   

   EstimatedSalary  Gender_Female  Gender_Male  Geography_France  \
0           101348              1            0                 1   
1           112542              1            0                 0   
2           113931              1            0                 1   
3            93826              1            0                 1   
4            79084              1            0                 0   

   Geography_Germany  Geography_Spain  
0                  0                0  
1                  0                1  
2                  0                0  
3    

In [ ]:
scaler = MinMaxScaler() # TODO: Lakukan scaling atau normalisasi
X_transform = scaler.fit_transform(X)

print(X_transform[:5])

[[0.32432432 0.2        0.         0.         1.         1.
  0.50673314 1.         0.         1.         0.         0.        ]
 [0.31081081 0.1        0.33402817 0.         0.         1.
  0.56270846 1.         0.         0.         0.         1.        ]
 [0.32432432 0.8        0.63635422 0.66666667 1.         0.
  0.56965412 1.         0.         1.         0.         0.        ]
 [0.28378378 0.1        0.         0.33333333 0.         0.
  0.46911957 1.         0.         1.         0.         0.        ]
 [0.33783784 0.2        0.50024313 0.         1.         1.
  0.39540256 1.         0.         0.         0.         1.        ]]


In [ ]:
# Bagian ini digunakan jika Anda melakukan scaling atau normalisasi.
# Jika tidak, code ini bisa dilewati dan diganti dengan code yang ada di dalam comment.

X_transform = pd.DataFrame(X_transform, columns = X.columns)
# X_transform = X.copy()

### Train-Test Split

In [ ]:
X_train ,X_test, y_train, y_test = tts(X_transform, y, test_size = 0.25, random_state = 123)


### Model #1

In [ ]:
# [ PERTANYAAN ]
# TODO: Jelaskan secara singkat model #1 yang Anda gunakan.


Model pertama yang digunakan adalah Logistic Regression yang biasanya digunakan untuk memprediksi probabilitas dari dua kelas. Cara model ini bekerja adalah dengan mengestimasi parameter dari fungsi logit (logaritma odds), sehingga output dari model ini berupa probabilitas yang diubah menjadi label klasifikasi menggunakan threshold tertentu.

In [ ]:
# CONTOH pemanggilan library dan penggunaannya:

# from sklearn.linear_model import LogisticRegression
# model1 = LogisticRegression()
# params = {"tol": [0.1,0.01,0.001], 'C':[0.5,1.0,1.5,2.0]}


# TODO: Pilih salah satu model Machine Learning (ML) dari Scikit-learn
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import GridSearchCV

model1 = LogisticRegression()
params = {"tol": [0.1, 0.01, 0.001], 'C': [0.5, 1.0, 1.5, 2.0]}

# OPTIONAL: Lakukan hyperparameter tuning sesuai kebutuhan
grid = GridSearchCV(
    estimator= model1,
    param_grid= params,
    scoring = 'accuracy',
    n_jobs = 10, # core cpu yang digunakan
    cv = 10 # 3-fold cross validation (artinya kita melakukan iterasi model sebanyak 3 kali)
)

grid.fit(X_train,y_train)
grid.best_params_

{'C': 0.5, 'tol': 0.001}

In [ ]:
y_pred = y_pred = grid.best_estimator_.predict(X_test) # TODO: Lakukan evaluasi untuk model #1

print("Classification Report:")
print(classification_report(y_test,y_pred))
print("")

print("Confusion Matrix:")
print(confusion_matrix(y_test,y_pred))
print("")

print("Accuracy Score:")
print(accuracy_score(y_test, y_pred))

Classification Report:
              precision    recall  f1-score   support

           0       0.82      0.97      0.89      1983
           1       0.66      0.20      0.31       517

    accuracy                           0.81      2500
   macro avg       0.74      0.59      0.60      2500
weighted avg       0.79      0.81      0.77      2500


Confusion Matrix:
[[1930   53]
 [ 413  104]]

Accuracy Score:
0.8136


### Model #2

In [ ]:
# [ PERTANYAAN ]
# TODO: Jelaskan secara singkat model #2 yang Anda gunakan.

Untuk model kedua, disini digunakan Random Forest. Random forest bisa dikatakan sebagai algoritma ensemble yang terdiri dari banyak pohon keputusan independen yang dilatih pada subset acak dari data. Cara kerja dari random forest sendiri adalah dengan membuat banyak decision tree yang kemudian digabungkan hasilnya untuk membuat sebuah prediksi.

In [ ]:
# CONTOH pemanggilan library dan penggunaannya:

# from sklearn.linear_model import LogisticRegression
# model1 = LogisticRegression()
# params = {"tol": [0.1,0.01,0.001], 'C':[0.5,1.0,1.5,2.0]}


# TODO: Pilih salah satu model Machine Learning (ML) dari Scikit-learn
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import GridSearchCV

model2 = RandomForestClassifier()
params = {
    'n_estimators': [50, 100],
    'max_depth': [10, None],
    'min_samples_split': [2, 5],
    'min_samples_leaf': [1, 2],
    'bootstrap': [True]
}

# OPTIONAL: Lakukan hyperparameter tuning sesuai kebutuhan
grid = GridSearchCV(
    estimator= model2,
    param_grid= params,
    scoring = 'accuracy',
    n_jobs = 10, # core cpu yang digunakan
    cv = 10 # 3-fold cross validation (artinya kita melakukan iterasi model sebanyak 3 kali)
)

grid.fit(X_train,y_train)
grid.best_params_

{'bootstrap': True,
 'max_depth': 10,
 'min_samples_leaf': 1,
 'min_samples_split': 5,
 'n_estimators': 50}

In [ ]:
y_pred = y_pred = grid.best_estimator_.predict(X_test) # TODO: Lakukan evaluasi untuk model #2

print("Classification Report:")
print(classification_report(y_test,y_pred))
print("")

print("Confusion Matrix:")
print(confusion_matrix(y_test,y_pred))
print("")

print("Accuracy Score:")
print(accuracy_score(y_test, y_pred))

Classification Report:
              precision    recall  f1-score   support

           0       0.87      0.97      0.92      1983
           1       0.81      0.44      0.57       517

    accuracy                           0.86      2500
   macro avg       0.84      0.70      0.74      2500
weighted avg       0.86      0.86      0.85      2500


Confusion Matrix:
[[1929   54]
 [ 291  226]]

Accuracy Score:
0.862


### Model #3

In [ ]:
# [ PERTANYAAN ]
# TODO: Jelaskan secara singkat model #3 yang Anda gunakan.

SVM merupakan algoritma yang mencari hyperplane optimal untuk memisahkan kelas-kelas dalam ruang fitur. Untuk cara kerjanya, SVM mencari margin terbesar antara kelas-kelas, dan dapat menggunakan kernel trick untuk memetakan data ke ruang berdimensi lebih tinggi untuk menangani masalah klasifikasi non-linear.

In [ ]:
# CONTOH pemanggilan library dan penggunaannya:

# from sklearn.linear_model import LogisticRegression
# model1 = LogisticRegression()
# params = {"tol": [0.1,0.01,0.001], 'C':[0.5,1.0,1.5,2.0]}


# TODO: Pilih salah satu model Machine Learning (ML) dari Scikit-learn
from sklearn.svm import SVC
from sklearn.model_selection import GridSearchCV

model3 = SVC()
params = {
    'C': [0.1, 1, 10,100],
    'gamma': ['scale', 'auto'],
    'kernel': ['linear', 'rbf', 'poly']
}

# OPTIONAL: Lakukan hyperparameter tuning sesuai kebutuhan
grid = GridSearchCV(
    estimator= model3,
    param_grid= params,
    scoring = 'accuracy',
    n_jobs = 10, # core cpu yang digunakan
    cv = 10 # 3-fold cross validation (artinya kita melakukan iterasi model sebanyak 3 kali)
)

grid.fit(X_train,y_train)
grid.best_params_

{'C': 100, 'gamma': 'scale', 'kernel': 'poly'}

In [ ]:
y_pred = y_pred = grid.best_estimator_.predict(X_test) # TODO: Lakukan evaluasi untuk model #3

print("Classification Report:")
print(classification_report(y_test,y_pred))
print("")

print("Confusion Matrix:")
print(confusion_matrix(y_test,y_pred))
print("")

print("Accuracy Score:")
print(accuracy_score(y_test, y_pred))

Classification Report:
              precision    recall  f1-score   support

           0       0.87      0.98      0.92      1983
           1       0.82      0.44      0.57       517

    accuracy                           0.86      2500
   macro avg       0.85      0.71      0.75      2500
weighted avg       0.86      0.86      0.85      2500


Confusion Matrix:
[[1934   49]
 [ 289  228]]

Accuracy Score:
0.8648


### Conclusion

In [ ]:
# [ PERTANYAAN ]
# TODO: Tarik kesimpulan model mana yang terbaik beserta alasannya

Dari hasil tes ketiga model, SVM merupakan model yang paling baik untuk digunakan pada kasus ini dikarenakan beberapa hal berikut :

1. Akurasi Tertinggi: SVM memiliki akurasi kurang lebih sebesar 0.8648, yang merupakan nilai tertinggi dibandingkan dengan Random Forest (kurang lebih 0.8596) dan Logistic Regression (kurang lebih 0.8136). Ini menunjukkan bahwa secara keseluruhan, SVM lebih sering menghasilkan prediksi yang benar.

2. Keseimbangan Precision dan Recall: SVM menawarkan keseimbangan terbaik antara precision dan recall, terutama pada Class 1 (Exited). Meskipun recall Class 1 pada SVM dan Random Forest sama (0.44), precision SVM (0.82) lebih baik daripada Random Forest (0.79), yang berarti SVM lebih andal dalam menghindari kesalahan deteksi Class 1 yang tidak benar.

3. F1-Score: SVM memiliki F1-score yang lebih tinggi (0.57) untuk Class 1, yang mencerminkan kombinasi precision dan recall yang lebih baik. Hal ini menjadikan SVM lebih mampu menangani imbalance pada data, di mana jumlah Class 0 jauh lebih banyak daripada Class 1.

4. Confusion Matrix: SVM menghasilkan lebih sedikit kesalahan pada Class 0 dibandingkan dengan Random Forest, dan kesalahan pada Class 1 juga sedikit lebih baik dari Random Forest.

Jadi, bisa disimpulkan bahwa SVM memberikan performa yang jauh lebih baik dan konsisten dibanding dengan 2 model lain yang digunakan pada kasus ini.



### Scoring
Total `#TODO` = 13
<br>Checklist:

- [ ] Hilangkan kolom yang tidak relevan untuk pemodelan
- [ ] Lakukan One-Hot Encoder untuk data kategorikal, dengan fungsi pd.get_dummies
- [ ] Pisahkan fitur (X) dengan target (Y), dimana Y merujuk ke kolom "Exited"
- [ ] Lakukan scaling atau normalisasi
- [ ] Jelaskan secara singkat model #1 yang Anda gunakan
- [ ] Pilih salah satu model Machine Learning (ML) dari Scikit-learn (model #1)
- [ ] Lakukan evaluasi untuk model #1
- [ ] Jelaskan secara singkat model #2 yang Anda gunakan
- [ ] Pilih salah satu model Machine Learning (ML) dari Scikit-learn (model #2)
- [ ] Lakukan evaluasi untuk model #2
- [ ] Jelaskan secara singkat model #3 yang Anda gunakan
- [ ] Pilih salah satu model Machine Learning (ML) dari Scikit-learn (model #3)
- [ ] Lakukan evaluasi untuk model #3

### Additional readings
- N/A

### Copyright © 2024 Startup Campus, Indonesia
* You may **NOT** use this file except there is written permission from PT. Kampus Merdeka Belajar (Startup Campus).
* Please address your questions to mentors.